apps/web-staking/src/services/redemptions.service.ts

```diff
@@ -0,0 +1,230 @@
+				claimable.push(redemption);
+			} else {
+				redemption.endTime = redemption.startTime + redemption.duration;
+				open.push(redemption);
+			}
+		}
+	}
+	return {
+		claimable: sortLists(claimable),
+		open: sortLists(open),
+		closed: sortLists(closed, 'desc')
+	};
+};
+
+async function getAllUserRedemptionsFromChain(web3Instance: Web3Instance, walletAddress: string): Promise<RedemptionRequest[]> {
+	const allRedemptions: RedemptionRequest[] = [];
+
+	let totalUserRedemptions: number = -1;
+	let offset = 0;
+
+	while (totalUserRedemptions < 0 || totalUserRedemptions > offset) {
+		let { redemptions, totalRedemptions } = await callGetAllRedemptionsFunctionOnContract(web3Instance, walletAddress, REDEMPTION_BATCH_SIZE, offset);
+		allRedemptions.push(...redemptions);
+		offset += REDEMPTION_BATCH_SIZE;
+		totalUserRedemptions = totalRedemptions;
+	}
+	return allRedemptions;
+}
+
+function extractIdsToBatches(redemptions: RedemptionRequest[]): number[][] {
+	const batches: number[][] = [];
+	if (redemptions.length > REDEMPTION_BATCH_SIZE) {
+		const totalBatches = Math.ceil(redemptions.length / REDEMPTION_BATCH_SIZE);
+		for (let i = 0; i < totalBatches; i++) {
+			const start = i * REDEMPTION_BATCH_SIZE;
+			const end = start + REDEMPTION_BATCH_SIZE;
+			batches.push(redemptions.slice(start, end).map(r => r.index));
+		}
+		return batches;
+	}
+
+	const singleBatch = redemptions.map(r => r.index);
+	batches.push(singleBatch);
+
+	return batches;
+}
+
+async function callRefreshRedemptionsFunctionOnContract(web3Instance: Web3Instance, walletAddress: string, indices: number[]): Promise<{ redemptions: RedemptionRequest[], totalRedemptions: number }> {
+	try {
+		const esXaiContract = new web3Instance.web3.eth.Contract(esXaiAbi, web3Instance.esXaiAddress);
+		const result = await esXaiContract.methods.refreshUserRedemptionsByIndex(walletAddress, indices).call() as unknown[];
+		const redemptions = result[0] as unknown[] as RedemptionRequest[];
+		const totalRedemptions = Number(result[1]);
+		return { redemptions, totalRedemptions };
+	} catch (error) {
+		console.error("Error fetching refreshed redemptions from contract:", error);
+		return { redemptions: [], totalRedemptions: 0 };
+	}
+}
+
+async function callGetAllRedemptionsFunctionOnContract(web3Instance: Web3Instance, walletAddress: string, batchSize: number, offset: number): Promise<{ redemptions: RedemptionRequest[], totalRedemptions: number }> {
+	try {
+		const esXaiContract = new web3Instance.web3.eth.Contract(esXaiAbi, web3Instance.esXaiAddress);
+		const result = await esXaiContract.methods.getRedemptionsByUser(walletAddress, batchSize, offset).call() as unknown[];
+		const redemptions = result[0] as unknown[] as RedemptionRequest[];
+		const totalRedemptions = Number(result[1]);
+		return { redemptions, totalRedemptions };
+	} catch (error) {
+		console.error("Error fetching user redemptions from contract:", error);
+		return { redemptions: [], totalRedemptions: 0 };
+	}
+
+}
```

Comment:
These two look almost identical.

Reply:
We talked about this before, see screenshot. Probably could have been better, but for this story it was just to move the existing code. 
[Previous PR](https://github.com/xai-foundation/sentry/pull/406)
![Previous](https://github.com/user-attachments/assets/f19e97e6-c26d-4c18-8f81-8f901b8df275)